internal/aws/xray/telemetry/recorder_test.go

```diff
@@ -0,0 +1,194 @@
+	mc.On("PutTelemetryRecords", mock.Anything).Return(nil, nil).Once()
+	mc.On("PutTelemetryRecords", mock.Anything).Return(nil, errors.New("error"))
+	recorder := registry.Register(
+		component.NewID("test"),
+		mc,
+		nil,
+		&Config{IncludeMetadata: false},
+		nil,
+	).(*telemetryRecorder)
+	recorder.interval = 50 * time.Millisecond
+	recorder.Start()
+	defer recorder.Stop()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		ticker := time.NewTicker(time.Millisecond)
+		for {
+			select {
+			case <-ticker.C:
+				recorder.RecordSegmentsReceived(1)
+				recorder.RecordSegmentsSpillover(1)
+				recorder.RecordSegmentsRejected(1)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	assert.Eventually(t, func() bool {
+		return mc.count.Load() >= 2
+	}, time.Second, 5*time.Millisecond)
+}
+
+func TestIncludeMetadata(t *testing.T) {
+	sess := awsmock.Session
+	recorder := newTelemetryRecorder(
+		nil,
+		sess,
+		&Config{IncludeMetadata: true},
+		&awsutil.AWSSessionSettings{ResourceARN: "session_arn"},
+	).(*telemetryRecorder)
+	assert.Equal(t, "", recorder.hostname)
+	assert.Equal(t, "", recorder.instanceID)
+	assert.Equal(t, "session_arn", recorder.resourceARN)
+
+	t.Setenv(envAWSHostname, "env_hostname")
+	t.Setenv(envAWSInstanceID, "env_instance_id")
+	recorder = newTelemetryRecorder(
+		nil,
+		sess,
+		&Config{IncludeMetadata: true},
+		nil,
+	).(*telemetryRecorder)
+	assert.Equal(t, "env_hostname", recorder.hostname)
+	assert.Equal(t, "env_instance_id", recorder.instanceID)
+	assert.Equal(t, "", recorder.resourceARN)
+}
+
+func TestRecordConnectionError(t *testing.T) {
+	type testParameters struct {
+		errorCode       string
+		errorStatusCode int
+		field           func(connectionErrors *xray.BackendConnectionErrors) *int64
+	}
+	testCases := []testParameters{
+		{
+			errorStatusCode: http.StatusInternalServerError,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.HTTPCode5XXCount
+			},
+		},
+		{
+			errorStatusCode: http.StatusBadRequest,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.HTTPCode4XXCount
+			},
+		},
+		{
+			errorStatusCode: http.StatusFound,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.OtherCount
+			},
+		},
+		{
+			errorCode: request.ErrCodeResponseTimeout,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.TimeoutCount
+			},
+		},
+		{
+			errorCode: request.ErrCodeRequestError,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.UnknownHostCount
+			},
+		},
+		{
+			errorCode: request.ErrCodeSerialization,
+			field: func(connectionErrors *xray.BackendConnectionErrors) *int64 {
+				return connectionErrors.OtherCount
+			},
+		},
+	}
+	recorder := newTelemetryRecorder(
+		nil,
+		nil,
+		nil,
+		nil,
```

Comment:
I am not a big fan of having `nil`to represent a "safe" default, mostly because it is easy to trip or means there is some additional checks to see if the value is safe to use and some amount code duplication.

Is there a way you can either adopt an options pattern or use provide `nop` values?

Reply:
Adopted the options pattern and created some basic nops for recorder/registry.